service/demo.go

```diff
@@ -120,9 +113,90 @@ func (ts *demoService) GetUsers(dataFileName string) (*[]types.User, error) {
 	return &users, nil
 }
 
-func (ts *demoService) GetUsersMap(dataFileName string) (map[int]types.User, error) {
-	log.Println("In service | GetUsersMap")
+func (ts *demoService) GetUsersConcurrently(dataFileName string, idType string, items int, itemsPerWorker int) (*[]types.User, error) {
```

Comment:
Even though, you are using concurrency, you did not used the worker pool pattern, instead you just used the fan out pattern and then just fan in, this could seem the same but it is not, due to some things:

- A worker pool know how many workers it has and have control of the channels being used to pass the "jobs" functions as the one receiving the final data.
- A WaitGroup could probably worked better instead of playing with the I variable to know when you should get out of the for.

Another thing is that the method has a huge bottleneck, this is reading one record each time, one thing you could probably do differently is to read the file once, then pass through a channel the required data to process, and then finally receive the result in the result channel.

Reply:
```Another thing is that the method has a huge bottleneck, this is reading one record each time, one thing you could probably do differently is to read the file once, then pass through a channel the required data to process, and then finally receive the result in the result channel.```
if i need 10 records and have 1000 in my csv I should not read all file

Reply:
```A worker pool know how many workers it has and have control of the channels being used to pass the "jobs" functions as the one receiving the final data.```
need to investigate what is worker pool.
I did not see the record of the last workshop. Maybe you told about it there